### PR TITLE
[RFC][V1] `LogitsProcessor` interface

### PR DESCRIPTION
**Pull request by njhill**
Proposed abstraction for how to handle sampling parameters in relation to the persistent batch. This interface could then be used as an extension point for custom logits processors.

Key goals/ideas:
- Logits processor implementations are configured globally, we won't support per-request
- They apply at a batch level rather than per-request to allow for / encourage vectorized application
- Each logits processor encapsulates its own state and is responsible for updating it as needed based on notification of persistent batch updates and new output tokens each step. This minimizes the number of times tensors need to be reconstructed and updated on the GPU.

To demonstrate the idea I've implemented LPs for `min_tokens`, `logit_bias` and `min_p`, but if we decide to go this route it should be straightforward to refactor the others similarly.

Note this is just to discuss the general approach - it could still be simplified/refined further.

```python
class LogitsProcessor(ABC):
    @abstractmethod
    def apply(self, logits: torch.Tensor) -> torch.Tensor:
        raise NotImplementedError

    @abstractmethod
    def update_states(
        self,
        batch_update: Optional[BatchUpdate] = None,
    ) -> None:
        """Called when there are new output tokens, prior
        to each forward pass.
        Args:
            batch_update is non-None iff there have been
            changes to the batch makeup.
        """
        raise NotImplementedError
        
@dataclasses.dataclass
class BatchUpdate:
    # Batch indices of any removed requests.
    removed: List[int]
    # (from, to) batch indices of any requests
    # moved within the batch.
    moved: List[Tuple[int, int]]
    # (index, params, output_tok_ids) for new
    # requests added to the batch.
    #TODO may need to include one or two other things here, like prompt token ids.
    added: List[Tuple[int, SamplingParams, List[int]]]
    # The current number of requests in the batch.
    batch_size: int
```

@WoosukKwon @AlpinDale @houseroad

### CODE REVIEW

File: vllm/v1/sample/logits_processor.py [hunk 1]
```diff
@@ -0,0 +1,242 @@
...
+
+        self.min_p_cpu_tensor = torch.zeros((max_num_reqs, ),
+                                            dtype=torch.float32,
+                                            device="cpu",
+                                            pin_memory=pin_memory)
+        self.min_p_cpu = self.min_p_cpu_tensor.numpy()
+        # Pre-allocated device tensor
+        self.min_p_gpu: torch.Tensor = torch.empty((max_num_reqs, ),
+                                                   dtype=torch.float32,
+                                                   device=device)
```
Comment (WoosukKwon) [inline vllm/v1/sample/logits_processor.py:63]: Can we preallocate for other as well?

**Review by WoosukKwon — COMMENTED** [review]
Looks very good to me!

File: vllm/v1/sample/logits_processor.py (hunk 1 above)
Comment (njhill) [reply vllm/v1/sample/logits_processor.py:63]: @WoosukKwon this could potentially be done but it's not quite a simple because the maximum size could be much larger (e.g. `max_batch_size * max_logit_bias_tokens * 3` for logit bias), but they are sparse and so in practice the size would be much smaller, or more likely they wouldn't actually be used at all.

Intuitively, since we are minimizing how often these get updated, and still doing the transfer from cpu async from pinned mem, I would guess we aren't losing much by not preallocating. But I can benchmark and see. If we do it we may want to just allocate to high-water-mark so that if the feature isn't used for example then there will be no allocation.

Comment (WoosukKwon) [reply vllm/v1/sample/logits_processor.py:63]: Got it. Sounds good. As this is a niche feature, I think we shouldn't spend too much time in it.

File: vllm/v1/sample/logits_processor.py [hunk 2]
```diff
@@ -0,0 +1,242 @@
...
+                 device: torch.device):
+        self.min_p_count: int = 0
+
+        self.min_p_cpu_tensor = torch.zeros((max_num_reqs, ),
+                                            dtype=torch.float32,
+                                            device="cpu",
+                                            pin_memory=pin_memory)
+        self.min_p_cpu = self.min_p_cpu_tensor.numpy()
+        # Pre-allocated device tensor
+        self.min_p_gpu: torch.Tensor = torch.empty((max_num_reqs, ),
```
Comment (houseroad) [inline vllm/v1/sample/logits_processor.py:61]: can we call it min_p_device? We may have use case other than GPU, right?

File: vllm/v1/sample/logits_processor.py [hunk 3]
```diff
@@ -0,0 +1,242 @@
...
+
+    def __init__(self, max_num_reqs: int, pin_memory: bool,
+                 device: torch.device):
+        self.min_p_count: int = 0
+
+        self.min_p_cpu_tensor = torch.zeros((max_num_reqs, ),
+                                            dtype=torch.float32,
+                                            device="cpu",
+                                            pin_memory=pin_memory)
+        self.min_p_cpu = self.min_p_cpu_tensor.numpy()
```
Comment (houseroad) [inline vllm/v1/sample/logits_processor.py:59]: why having min_p_cpu on top of min_p_cpu_tensor?

File: vllm/v1/sample/metadata.py [hunk 4]
```diff
@@ -33,7 +34,5 @@ class SamplingMetadata:
 
     output_token_ids: List[List[int]]
 
-    # req_index -> (min_tokens, stop_token_ids)
-    min_tokens: Dict[int, Tuple[int, Set[int]]]
-
-    logit_bias: List[Optional[Dict[int, float]]]
+    logits_procs: List[LogitsProcessor]
+    nongreedy_logits_procs: List[LogitsProcessor]
```
Comment (houseroad) [inline vllm/v1/sample/metadata.py:38]: wondering why split into nongreedy and regular logits processors?

File: vllm/v1/sample/logits_processor.py (hunk 3 above)
Comment (njhill) [reply vllm/v1/sample/logits_processor.py:59]: This mirrors the existing implementation, I think the reason is that it's much cheaper to manipulate individual elements in numpy arrays than tensors, so better to do those updates on a numpy view. But we still need the tensor to transfer to GPU.

File: vllm/v1/sample/metadata.py (hunk 4 above)
Comment (njhill) [reply vllm/v1/sample/metadata.py:38]: Some logits processors don't affect greedy decoding (or if they do, only due to precision errors), so we want to only apply them when there are non-greedy (random-sampled) requests in the batch. I don't really like this name for the field  but couldn't think of anything better :)

In theory, each of the logits processors keeps track of whether there are any applicable requests in the current batch and so this might not be needed. But it means they would each need to check the temperature too when requests are added. AFAIK we don't have validation to reject requests that have e.g. temp=0 and top_p=0.5 (top_p has no effect in this case).

**Comment by mmoskal** [comment]
@njhill shouldn't BatchUpdate also include the tokens that were sampled for each current sequence? it's not needed for the min_p/min_tokens/logit_bias ones you implemented, but would be needed for anything more complicated?

**Comment by njhill** [comment]
> @njhill shouldn't BatchUpdate also include the tokens that were sampled for each current sequence? it's not needed for the min_p/min_tokens/logit_bias ones you implemented, but would be needed for anything more complicated?

So this is not necessarily the final state, but currently this it's handled via the list that's passed in the `added` requests. This is the list that is updated in-place with new tokens and so the impl can hang on to that if it needs to know them, and check it each time it's called. The idea is that BatchUpdate will only be present if reqs have been added or removed from the batch, otherwise this list of output ids can be checked for all of the current requests that the LP cares about. Of course this will need to be clearly documented if it remains the way that it's done :)

You can see an example of this for the [`min_tokens` impl](https://github.com/vllm-project/vllm/pull/13360/files#diff-d01f143e1af472f24af24842cb879907ce624e6e5c977935e944545240723529R207) in this PR, which uses the length of that list:

**Comment by jiwangyihao** [comment]
In the V1 interface, I'm curious if the following code can still be used to customize the Logits processor for a single batch?

```python
outputs = llm.generate(
    prompts,
    SamplingParams(
        temperature=temperature,
        max_tokens=1,
        allowed_token_ids=accepted_ids
    ),
)
```

And if not, is there any alternative solution?

**Comment by mmoskal** [comment]
You can set the guided decoding backend to "guidance" and use the following grammar, assuming your tokens are 74, 837, 1010, 1011, 1012, 1013:

```lark
start: ( <[74]> | <[837]> | <[1010-1013]> )*
```

for details see https://github.com/guidance-ai/llguidance/blob/main/docs/syntax.md#special-tokens

Make sure you have llguidance at least 0.7.10 (should be ready in 20 min); 0.7.9 as used in vllm would reject this grammar in front-end.

Of course you can also use strings in your grammar, not tokens, but then we will allow any tokenization of the given string, which may or may not be what you want.

**Comment by wlhgtc** [comment]
If I want to force stop thinking in an R1-like model (e.g., when `prompt + outputs > 8192`, force generate `</think>`)
how can I get `prompt_length` for single request?
Thank you for your time.

File: vllm/v1/sample/logits_processor.py [hunk 5]
```diff
@@ -0,0 +1,242 @@
...
+            self.min_p = self.min_p_gpu[:size]
+            self.min_p.copy_(self.min_p_cpu_tensor[:size], non_blocking=True)
+            self.min_p.unsqueeze_(1)
+
+    def apply(self, logits: torch.Tensor) -> torch.Tensor:
+        if not self.min_p_count:
+            return logits
+
+        # Convert logits to probability distribution
+        probability_values = torch.nn.functional.softmax(logits, dim=-1)
```
Comment (DreamGenX) [inline vllm/v1/sample/logits_processor.py:106]: You can do min p without softmax, which might be faster:
Set logits[i] to -inf when logits[i] < logits_max + log(min_p) where logits_max is the largest logit.

You can precompute the log(min_p) in update_states

**Comment by njhill** [comment]
I've just rebased this PR, but haven't yet addressed some changes made since it was originally created:
- The TPU impl has [separate sampling metadata](https://github.com/vllm-project/vllm/blob/main/vllm/v1/sample/tpu/metadata.py) handling, we need to see how this will work with that.
- There is a [check for whether spec decode is supported](https://github.com/vllm-project/vllm/blob/main/vllm/v1/spec_decode/utils.py#L5) based on the state of the input batch. Now that the input batch will change to contain LPs instead of the "raw" sampling param data, this check will need to be adjusted.
- A out-of-vocab token check [was added](https://github.com/vllm-project/vllm/pull/16529) to the logit bias impl in the main branch. The re-implemented logit bias logic in this PR doesn't include that, but we should move that check to be part of the request validation anyhow (as I've now commented on that PR).

**Review by NickLucche — COMMENTED** [review]
Nice work!
I assume no in-place update operations are going to be allowed in order to not enforce any priority on the order on which logits processor in `logits_procs` are applied right?

> The TPU impl has separate sampling metadata handling, we need to see how this will work with that.

The main thing with TPU is that we need a way to limit the logits processors to the params that are currently supported before tracing in a fixed order.  With the current implementation there's no way to toggle on/off some of them so we'll have to edit `logits_procs` from within `TPUModelRunner`. 
Also, we're currently replacing the `update_states` function with one that does not introduce op dynamism for tpus. Here the `update_states` is scattered across different logits processor so we would have to address each one individually.

As a rule of thumb, for TPU `update_states` functions are always ok when they operate on CPU tensors and then only move the result to device.  

cc @yaochengji

File: vllm/v1/worker/tpu_model_runner.py [hunk 6]
```diff
@@ -1059,13 +1059,13 @@ def _get_padded_num_reqs_with_upper_limit(x: int, upper_limit: int) -> int:
...
-    """Generate a list of padding size, starting from min_token_size, 
+    """Generate a list of padding size, starting from min_token_size,
     ending with a number that can cover max_token_size
-    
+
     If padding_gap == 0 then:
         increase 2X each time (exponential)
     else:
-        first increase the size to twice, 
+        first increase the size to twice,
```
[the remark this answers is not in the document]
Comment (njhill) [reply vllm/v1/worker/tpu_model_runner.py:1068]: Sorry these must have been inadvertent while rebasing, probably by IDE. We can revert them.

File: vllm/v1/attention/backends/mla/common.py [hunk 7]
```diff
@@ -415,20 +419,25 @@ def reorder_batch(self, input_batch: "InputBatch",
...
+            i1 = prefills[i - 1]
+            i2 = decodes[num_decodes - i]
+            input_batch.swap_states(i1, i2)
+
+            # Using "move" operation of LogitsProcessors via temporary slot
+            # currently.
+            # TODO possibly add more direct swap operation to LPs
+            swaps.append((i1, input_batch.max_num_reqs))
+            swaps.append((i2, i1))
+            swaps.append((input_batch.max_num_reqs, i2))
```
Comment (LucasWilkinson) [inline vllm/v1/attention/backends/mla/common.py:439]: we could maybe consider having `input_batch` track the swaps in an internal data structure that gets reset with calls to `refresh_sampling_metadata`, this way the if we do do `TODO possibly add more direct swap operation to LPs` then all the attn metadata implementations don't  need to updated (FlashInfer does swapping to)

**Comment by afeldman-nm** [comment]
> * was added

@njhill see my comment here

https://github.com/vllm-project/vllm/pull/16529#discussion_r2075990154

isn't the validation inside of the logit bias logits processor redundant?

**Comment by hmellor** [comment]
Closing — this RFC was implemented and merged via #16728.
